lib/gtfs.ex

```diff
@@ -73,16 +73,38 @@ defmodule Gtfs do
   end
 
   def handle_call({:timepoints_on_route, route_id}, _from, {:loaded, gtfs_data} = state) do
-    timepoint_ids =
+    route_patterns_by_direction =
       gtfs_data.route_patterns
       |> Enum.filter(fn route_pattern -> route_pattern.route_id == route_id end)
-      |> Enum.map(fn route_pattern -> route_pattern.representative_trip_id end)
-      |> Enum.flat_map(fn trip_id -> gtfs_data.stop_times[trip_id] end)
-      |> Enum.map(fn stop_time -> stop_time.timepoint_id end)
-      |> Enum.uniq()
-      |> List.delete("")
+      |> Enum.group_by(fn route_pattern -> route_pattern.direction_id end)
+
+    timepoints_by_direction =
+      Helpers.map_values(route_patterns_by_direction, fn route_patterns ->
+        timepoints_for_route_patterns(route_patterns, gtfs_data)
+      end)
 
-    {:reply, timepoint_ids, state}
+    merged_timepoint_ids =
+      Helpers.merge_lists([
+        Enum.reverse(Map.get(timepoints_by_direction, 0, [])),
+        Map.get(timepoints_by_direction, 1, [])
+      ])
+
+    {:reply, merged_timepoint_ids, state}
+  end
+
+  # All route_patterns should be in the same direction
+  @spec timepoints_for_route_patterns([RoutePattern.t()], t()) :: [Timepoint.id()]
+  defp timepoints_for_route_patterns(route_patterns, gtfs_data) do
+    route_patterns
+    |> Enum.map(fn route_pattern ->
+      stop_times = gtfs_data.stop_times[route_pattern.representative_trip_id]
+
+      stop_times
+      |> Enum.sort_by(fn stop_time -> stop_time.stop_sequence end)
+      |> Enum.map(fn stop_time -> stop_time.timepoint_id end)
+      |> Enum.filter(fn timepoint_id -> timepoint_id != nil end)
```

Comment:
It might be simpler to filter these out during the parsing stage.

Reply:
I've give that a shot.

Reply:
I've pushed it here: https://github.com/mbta/skate/commit/7e52f1fc67560ea77e8a0edfe56300de07f71aef

I'm not totally sold on it. It is less code, mostly because we don't need to store stop_times anymore. But it takes this file a step away from the simple model of "gtfs that we make queries against", and if we ever need the rest of the stop_times in the future, this will be net more complicated.

I'm leaning in favor of the preprocessing, but want a second opinion before including it in the PR.

Reply:
I don't think "gtfs that we make queries against" should be the model. GTFS is the source data, but when it's ingested, we should be processing it into structures that make it easy to query. The internal models should be designed to match the domain, not GTFS: it's why we should be using Timepoint.

An example from a different application is Concentrate, which has several servers each maintaining a view of part of the GTFS: https://github.com/mbta/concentrate/blob/master/lib/concentrate/filter/gtfs/supervisor.ex PickupDropOff is the one which looks at stop times if you want a more direct example.

Reply:
Okay. I'll merge it in.